### PR TITLE
Make notifications a polymorphic relation

### LINKED ISSUE #547: Decouple notification types from their target model
Currently we determine notification type based from its target model. Instead we should turn the notifications relation into a polymorphic one. This will make it possible to filter notifications by type for smiles and comment smiles again.

### PR DESCRIPTION
**Pull request by raccube**
Closes #547

### CODE REVIEW

File: db/migrate/20220720190421_add_type_to_notifications.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+class AddTypeToNotifications < ActiveRecord::Migration[6.1]
+  def up
+    add_column :notifications, :type, :string
```
Comment (nilsding) [inline db/migrate/20220720190421_add_type_to_notifications.rb:5]: Should probably be not nullable and have an index if you're doing queries based on this column.

File: db/migrate/20220720190421_add_type_to_notifications.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+    execute "UPDATE notifications SET type = 'Notification::Commented' WHERE target_type = 'Comment'"
+    execute "UPDATE notifications SET type = 'Notification::QuestionAnswered' WHERE target_type = 'Answer'"
+    execute "UPDATE notifications SET type = 'Notification::StartedFollowing' WHERE target_type = 'Relationship'"
+    execute <<~SQUIRREL
```
Comment (nilsding) [inline db/migrate/20220720190421_add_type_to_notifications.rb:11]: Oachkatzl!

File: app/models/notification.rb [hunk 3]
```diff
@@ -8,12 +10,12 @@ class << self
...
 
-    def for(recipient, options={})
-      self.where(options.merge!(recipient: recipient)).order(:created_at).reverse_order
+    def for(recipient, options = {})
+      where(options.merge!(recipient: recipient)).order(:created_at).reverse_order
     end
 
-    def for_type(recipient, type, options={})
-      self.where(options.merge!(recipient: recipient)).where('LOWER(target_type) = ?', type).order(:created_at).reverse_order
+    def for_type(recipient, type, options = {})
```
Comment (nilsding) [inline app/models/notification.rb]: the modern Ruby thing to do would be:
```suggestion
    def for_type(recipient, type, **kwargs)
```

`kwargs` will be your usual Hash

**Review by nilsding — CHANGES_REQUESTED** [review]

File: db/migrate/20220720190421_add_type_to_notifications.rb (hunk 1 above)
Comment (raccube) [reply db/migrate/20220720190421_add_type_to_notifications.rb:5]: 9103340c

File: app/models/notification.rb (hunk 3 above)
Comment (raccube) [reply app/models/notification.rb]: b1380922

File: db/migrate/20220720190421_add_type_to_notifications.rb (hunk 2 above)
Comment (raccube) [reply db/migrate/20220720190421_add_type_to_notifications.rb:11]: :shipit:

**Review by nilsding — APPROVED** [review]

**Pull request merged**